File: src/main/kotlin/com/github/tobi812/sprykerplugin/models/writer/FileWriter.java
package com.github.tobi812.sprykerplugin.models.writer;

import com.intellij.openapi.application.Result;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.fileEditor.OpenFileDescriptor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.*;
import com.jetbrains.php.lang.PhpFileType;
import org.jetbrains.annotations.NotNull;
import pav.sprykerFileCreator.model.ModelFactory;
import pav.sprykerFileCreator.model.generator.SprykerConstants;
import pav.sprykerFileCreator.model.manager.ClassManagerInterface;
import pav.sprykerFileCreator.model.matcher.ClassTypeMatcher;

import java.util.HashMap;

public class FileWriter implements FileWriterInterface {

    private  Project project;

    public FileWriter(Project project) {
        this.project = project;
    }

    @Override
    public PsiElement writeFile(String filePath, String fileName, String phpClassContent) throws Exception {
        PsiDirectory fileDirectory = this.getFileDirectory(filePath);
        PsiFile existingFile = fileDirectory.findFile(fileName + ".php");

        if (existingFile == null) {
            throw new Exception(fileName + " does already exist.");
        }

        PsiFile file = PsiFileFactory
                .getInstance(project)
                .createFileFromText(fileName + ".php", PhpFileType.INSTANCE, phpClassContent);

        new WriteCommandAction(this.project) {

            @Override
            protected void run(@NotNull Result result) throws Throwable {
                fileDirectory.add(file);

                PsiFile virtualFile = fileDirectory.findFile(fileName + ".php");
                if (virtualFile != null) {
                    new OpenFileDescriptor(project, virtualFile.getVirtualFile()).navigate(true);
                }
            }

            @Override
            public String getGroupID() {
                return "Create Command";
            }

        }.execute();

        return file;
    }

    private PsiDirectory getFileDirectory(String filePath) {

        PsiDirectory currentDir = this.getPsiDirectory();

        String[] pathSegments = filePath.split("/");

        for (String pathSegment : pathSegments) {
            PsiDirectory subDirectory = currentDir.findSubdirectory(pathSegment);

            if (subDirectory == null) {
                subDirectory = currentDir.createSubdirectory(pathSegment);
            }

            currentDir = subDirectory;
        }

        return currentDir;
    }

    private PsiDirectory getPsiDirectory() {
        VirtualFile baseDir = this.project.getBaseDir();

        return PsiManager.getInstance(this.project).findDirectory(baseDir);
    }

}

File: src/main/kotlin/com/github/tobi812/sprykerplugin/models/writer/FileWriter.kt
package com.github.tobi812.sprykerplugin.models.writer

import com.intellij.openapi.application.Result
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.fileEditor.OpenFileDescriptor
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.PlatformVirtualFileManager
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.psi.*
import com.jetbrains.php.lang.PhpFileType

import java.lang.Exception
import java.nio.file.Path

class FileWriter(private val project: Project) : FileWriterInterface {
    @Throws(Exception::class)
    override fun writeFile(filePath: String, fileName: String, phpClassContent: String): PsiElement {
        val fileDirectory = getFileDirectory(filePath)
        val existingFile = fileDirectory.findFile("$fileName.php")
            ?: throw Exception("$fileName does already exist.")
        val file = PsiFileFactory
            .getInstance(this.project)
            .createFileFromText("$fileName.php", PhpFileType.INSTANCE, phpClassContent)

        object : WriteCommandAction<Any?>(this.project) {
            @Throws(Throwable::class)
            override fun run(result: Result<*>) {
                fileDirectory.add(file)
                val virtualFile = fileDirectory.findFile("$fileName.php")
                if (virtualFile != null) {
                    OpenFileDescriptor(this.project, virtualFile.virtualFile).navigate(true)
                }
            }

            override fun getGroupID(): String {
                return "Create Command"
            }
        }.execute()
        return file
    }

    private fun getFileDirectory(filePath: String): PsiDirectory {
        var currentDir = this.getPsiDirectory()
        val pathSegments = filePath.split("/".toRegex()).toTypedArray()
        for (pathSegment in pathSegments) {
            var subDirectory = currentDir!!.findSubdirectory(pathSegment)
            if (subDirectory == null) {
                subDirectory = currentDir.createSubdirectory(pathSegment)
            }

            currentDir = subDirectory
        }

        return currentDir
    }

    private fun getPsiDirectory(): PsiDirectory? {
        val basePath: String? = this.project.basePath

        val baseDir = PlatformVirtualFileManager.getInstance().findFileByNioPath(basePath)

        return PsiManager.getInstance(this.project).findDirectory(baseDir)
    }
}